### ISSUE TITLE
[question] Should options affect private dependencies?

### ISSUE THREAD

**Issue by db4** [conan-io/conan#12213]
Suppose I have PkgA/1.0 that depends on PkgB/2.0 and PkgC/1.0. PkgC/1.0 in turn privately depends on PkgB/1.0. Now if I set `-o PkgB:option=value` while creating PkgA, both PkgB/2.0 and private PkgB/1.0 are affected. Is it expected? For me it makes private dependencies almost unusable.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Yes, if you define ``-o pkgb:option=value`` in the command line (recall, try to use lowercase packages always, as Conan 2.0 will force it), then yes, it is a global, pattern based match, so it will match all packages named with ``pkgb``. You can try to limit it to include the version, like ``-o pkgb/2.0*:option=value`` instead.

Another possibility is that only ``pkgc`` defines the specific option value that is different than the default, in its own recipe. After all, if it is private, the key is that it is only known by it, and not the downstream consumers.

**Comment by db4** [comment]
Hi @memsharded!

> Yes, if you define `-o pkgb:option=value` in the command line (recall, try to use lowercase packages always, as Conan 2.0 will force it)

Even for my internal packages? That's unfortunate, historically I use capitalized names..

> then yes, it is a global, pattern based match, so it will match all packages named with `pkgb`. You can try to limit it to include the version, like `-o pkgb/2.0*:option=value` instead.

Great, didn't know that. Thanks a lot!

> Another possibility is that only `pkgc` defines the specific option value that is different than the default, in its own recipe. After all, if it is private, the key is that it is only known by it, and not the downstream consumers.

It does, but PkgA needs to set an option for PkgB/2.0, that doesn't exist in PkgB/1.0. So I have
```
ERROR: PkgB/1.0: option 'some_option' doesn't exist
Possible options are [...
```

Unfortunately there is another issue with private dependencies:
```
ERROR: At least two recipes provides the same functionality:
 - 'libjpeg' provided by 'libjpeg-turbo/2.1.2', 'libjpeg/9d'
 ```
 libjpeg-turbo/2.1.2 is required normally, libjpeg/9d - privately. Is it possible to work around this?

**Comment by memsharded** [comment]
> Even for my internal packages? That's unfortunate, historically I use capitalized names..

Yes, this was approved by the Tribe 2.0, like 1,5 years ago: https://github.com/conan-io/tribe/pull/17. Capitalized names are problematic in many OS, and also a source of many confusions in the ecosystem as a whole, it was a mistake to allow them in 1.X. Conan 2.0 will allow a temporary workaround to enable the migration, but it is recommended to start using lowercase for packages asap.

You want to keep an eye on https://docs.conan.io/en/latest/conan_v2.html for other coming changes.

> It does, but PkgA needs to set an option for PkgB/2.0, that doesn't exist in PkgB/1.0. So I have

If you use ``*`` in patterns like ``pkga*:option=value``, it will not fail, because it is not considered an exact match

> 'libjpeg' provided by 'libjpeg-turbo/2.1.2', 'libjpeg/9d'

Yes, apparently the ``provides`` feature is a global one, the only possibility would be to remove one of the ``provides`` definition in the recipes. It might be worth to check this as a potential feature for Conan 2.0.

**Comment by db4** [comment]
> > 'libjpeg' provided by 'libjpeg-turbo/2.1.2', 'libjpeg/9d'
> 
> Yes, apparently the `provides` feature is a global one, the only possibility would be to remove one of the `provides` definition in the recipes. It might be worth to check this as a potential feature for Conan 2.0.

libjpeg/libjpeg-turbo are CCI recipes, I can't modify them. And they are used indirectly via different OpenCV packages (one normal, one private). I used build_requirements instead of private requirements before, but 1) such use is strongly discouraged by Conan team 2) override is not possible for build_requirements. So no acceptable solution yet to have a completely isolated static dependency? At least for Conan 1.x?

**Comment by memsharded** [comment]
> libjpeg/libjpeg-turbo are CCI recipes, I can't modify them. And they are used indirectly via different OpenCV packages (one normal, one private). 

It is still possible, requires some work, but possible:
- Clone conan-center-index-repo
- Modify the recipe
- conan create . version@

Then you can upload it together with your packages to your own repo, and make it the first remote. It will have priority and Conan will resolve against it instead the ConanCenter one.


> So no acceptable solution yet to have a completely isolated static dependency? At least for Conan 1.x?

Not for the ``provides`` feature. ``private`` requirements are complicated in 1.X, and are documented as such. I will have a look to make sure it will work better in 2.0.

**Comment by memsharded** [comment]
> Not for the provides feature. private requirements are complicated in 1.X, and are documented as such. I will have a look to make sure it will work better in 2.0.

I have checked it, this is already solved and tested in Conan 2.0, private (``visible=False`` trait in requires) requirements will not create ``provides`` conflicts. Closing this as solved in beta
